server/index.js

```diff
@@ -16,62 +15,74 @@ app.use(bodyParser.json());
 app.get('/traits/:product_id', (req, res) => {
   const id = req.params.product_id;
   const traits = fetchers.fetchTraitsForProduct(id);
-  traits
-    .then((results) => {
-      const traitProducts = results.traits.reduce((acc, trait) => {
-        return acc.concat(fetchers.fetchProductsForTrait(trait, id));
-      }, []);
-      Promise.all(traitProducts).then((resultsFinal) => {
+  traits.then((traitsData) => {
+    const traitProducts = traitsData.traits.reduce((acc, trait) => {
+      return acc.concat(fetchers.fetchProductsForTrait(trait, id));
+    }, []);
+    Promise.all(traitProducts)
+      .then((resultsFinal) => {
         resultsFinal.forEach((result) => {
           console.log('results', result.products);
           // if (result.products.indexOf(req.params.product_id) >= 0) {
           //   console.log('duplicate!');
           // }
           while (result.products.length < 4) {
-            let filler = Math.ceil(Math.random() * 100);
-            if (!result.products.includes(filler) && !result.products.includes(id)) {
+            const filler = Math.ceil(Math.random() * 100);
+            if (!result.products.includes(filler) && filler !== id) {
               result.products.push(filler);
             }
           }
+          console.log('resultsFinal', resultsFinal);
         });
-        res.set({ 'Access-Control-Allow-Origin': '*' });
-        console.log(resultsFinal);
-        res.send(JSON.stringify(resultsFinal));
-        const imagesNeeded = resultsFinal.reduce((acc, result) => {
-          return acc.concat(result.products);
-        }, []);
-        console.log('products to Micko', imagesNeeded);
-        const requestArray = encodeURI(JSON.stringify(imagesNeeded));
-        // const requestURL = `http://127.0.0.1:3000/api/${requestArray}?type=thumbnail`;
-        // console.log(requestURL);
-        // axios
-        //   .get(requestURL)
-        //   .then((results) => {
-        //     console.log(results.body);
-        //   })
-        //   .catch((err) => {
-        //     if (err) {
-        //       res.status(505).send('Please try again');
-        //     }
-        //   });
-        // res.send(traitProducts);
+        axios
+          .all(
+            resultsFinal.map((result) => {
+              const requestArray = encodeURI(JSON.stringify(result.products));
+              const requestURL = `http://127.0.0.1:3000/api/${requestArray}?type=thumbnail`;
```

Comment:
You might want to move the base URL to your .env file i.e. `IMG_API=http://127.0.0.1:3000/` so that when it is deployed it is easy to swap out anywhere it is used.

Reply:
oh yes good point